server/routes/charts.js

```diff
@@ -39,9 +40,45 @@ router.route('/charts/new')
 router.route('/charts/:chart_id')
   .get(ensureAuthenticated, async(req,res) => {
     try {
+      const data = []
+      const { dataDb, appDb } = req.app.locals
       const { chart_id } = req.params
+      const { access } =  await appDb
+        .collection(collections.users)
+        .findOne(
+          { uid: req.user },
+          { _id: 0, access: 1 }
+        )
+      const { assessment, variable, fieldLabelValueMap, title } = await dataDb
+        .collection(collections.charts)
+        .findOne({ _id: new ObjectID(chart_id) })
+
+      for await (const field of fieldLabelValueMap) {
+        const valueCountData = []
+
+        for await (const study of access ){
+          let count = 0;
+          const { label, value } = field;
+          const subjectList = await dataDb
+            .collection(collections.toc)
+            .find({ assessment, study }, { projection: { collection: 1, subject: 1, study: 1, _id: 0 }})
+            .map(doc => ({ ...doc, variable }))
+            .toArray()
+
+            for await (const doc of subjectList) {
+              const { collection, study, variable, subject } = doc
+              const counted = await dataDb
+                .collection(collection)
+                .findOne({ site: study, [variable]: value, subject_id: subject }, { projection: { [variable]: 1, _id: 0 }})
+              count = counted ? count+=1 : count
+            }
```

Comment:
I _think_ what this is doing is counting up the number of occurrences of a variable value. Would it be possible to do 1 query to get a count instead of a query for each subject? Something like

```js
const count = await dataDb
  .collection(collection)
  .countDocuments({ site: study, [variable]: value })
```

Reply:
[3d5d8c](https://github.com/AMP-SCZ/dpdash/pull/91/commits/3d5d8c2367dffb4b84223b17754ce6f1a2ecb5c6)

Reply:
the change you linked too looks like a really complex query. Conceptually the requirement doesn't seem very complex (get the count of subjects that have this value for this variable). I'm also not very familiar with Mongo so I think an explanation of the complexity could be helpful.

Reply:
Yes, you're right an explanation could help. What I've done is taken the queries I originally had and then converted them into an aggregation.

An aggregation in mongo takes in a pipeline, and a pipeline is made up of different stages, the stages take an input of documents and they perform queries, transformations or calculations.

A pipeline is an array a stage is an object, so a pipeline is an array of objects. To query, transform data etc... the aggregation pipeline has different operations which start with the $ sign
The ones are used are the following. 

`$match` this is the query in an the aggregate
`$project` this specifies the fields I want in my document        
 `$lookup` this allows me to join another collection by matching a foreign field shared by both collections. Within a lookup we can implement a pipeline on the joined collection.
`$unwind` it deconstructs an array field and outputs a new document for each of these fields